src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+            #[inline]
+            pub fn from_le_bytes(bytes: [u8; 32]) -> $ty { $ty(Uint256::from_le_bytes((bytes))) }
+
+            #[doc = concat!("Creates a [`", stringify!($ty), "`] from a big-endian byte slice.")]
+            #[inline]
+            pub fn from_be_slice(bytes: &[u8]) -> Result<$ty, ParseLengthError> {
+                let u = Uint256::from_be_slice(bytes)?;
+                Ok($ty(u))
+            }
+
+            #[doc = concat!("Creates a [`", stringify!($ty), "`] from a little-endian byte slice.")]
+            #[inline]
+            pub fn from_le_slice(bytes: &[u8]) -> Result<$ty, ParseLengthError> {
+                let u = Uint256::from_le_slice(bytes)?;
+                Ok($ty(u))
+            }
+
+            /// Converts `self` to a big-endian array.
+            ///
+            /// # Examples
+            ///
+            /// ```
+            #[doc = concat!("use bitcoin::", stringify!($ty), ";")]
+            #[doc = concat!("let a = ", stringify!($ty), "::from(0xdeadbeef);")]
+            /// assert_eq!(a.to_be(), [0, 0, 0, 0xdeadbeef]);
+            /// ```
+            #[inline]
+            pub fn to_be(self) -> [u64; 4] { self.0.to_be() }
+
+            /// Converts `self` to a little-endian array.
+            ///
+            /// # Examples
+            ///
+            /// ```
+            #[doc = concat!("use bitcoin::", stringify!($ty), ";")]
+            #[doc = concat!("let a = ", stringify!($ty), "::from(0xdeadbeef);")]
+            /// assert_eq!(a.to_le(), [0xdeadbeef, 0, 0, 0]);
+            /// ```
+            #[inline]
+            pub fn to_le(self) -> [u64; 4] { self.0.to_le() }
+
+            /// Converts `self` to a big-endian byte array.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+
+            /// Converts `self` to a little-endian byte array.
+            #[inline]
+            pub fn to_le_bytes(self) -> [u8; 32] { self.0.to_le_bytes() }
+        }
+
+        impl From<u64> for $ty {
+            #[inline]
+            fn from(x: u64) -> Self { $ty(Uint256::from(x)) }
+        }
+
+        impl fmt::Display for $ty {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
+        }
+
+        impl Encodable for $ty {
+            #[inline]
+            fn consensus_encode<W: crate::io::Write + ?Sized>(
+                &self,
+                w: &mut W,
+            ) -> Result<usize, crate::io::Error> {
+                self.0.consensus_encode(w)
+            }
+        }
+
+        impl Decodable for $ty {
+            fn consensus_decode<R: crate::io::Read + ?Sized>(
+                r: &mut R,
+            ) -> Result<$ty, crate::consensus::encode::Error> {
+                let u: Uint256 = Decodable::consensus_decode(r)?;
+                Ok($ty(u))
+            }
+        }
+    };
+}
+
+/// Implements `ops::Mul` by integer type in both directions. Requires `u32::from($int)`.
```

Comment:
Why it's restricted to `u32`? Would be nice to allow `u128`

Reply:
Mentioned elsewhere but flagging here as well, I did all integer sizes upto `u64`. The reason being that I changed the internal representation to us `[u128; 2]` and we can multiply two `u64`s inside of a `u128`.